lib/internal/webstreams/compression.js

```diff
@@ -0,0 +1,164 @@
+  return typeof value?.[kHandle] === 'object' &&
+         value?.[kType] === 'CompressionStream';
+}
+
+function isDecompressionStream(value) {
+  return typeof value?.[kHandle] === 'object' &&
+         value?.[kType] === 'DecompressionStream';
+}
+
+class CompressionStream {
+  /**
+   * @param {'deflate'|'gzip'} format
+   */
+  constructor(format) {
+    this[kType] = 'CompressionStream';
+    switch (format) {
+      case 'deflate':
+        this[kHandle] = createDeflate();
+        break;
+      case 'gzip':
+        this[kHandle] = createGzip();
+        break;
+      default:
+        throw new ERR_INVALID_ARG_VALUE('format', format);
+    }
+    this[kTransform] = newReadableWritablePairFromDuplex(this[kHandle]);
+  }
+
+  /**
+   * @readonly
+   * @type {ReadableStream}
+   */
+  get readable() {
+    if (!isCompressionStream(this))
+      throw new ERR_INVALID_THIS('CompressionStream');
+    return this[kTransform].readable;
+  }
+
+  /**
+   * @readonly
+   * @type {WritableStream}
+   */
+  get writable() {
+    if (!isCompressionStream(this))
+      throw new ERR_INVALID_THIS('CompressionStream');
+    return this[kTransform].writable;
+  }
+
+  [kInspect](depth, options) {
+    if (!isCompressionStream(this))
+      throw new ERR_INVALID_THIS('CompressionStream');
+    customInspect(depth, options, 'CompressionStream', {
+      readable: this[kTransform].readable,
+      writable: this[kTransform].writable,
+    });
+  }
+}
+
+class DecompressionStream {
+  /**
+   * @param {'deflate'|'gzip'} format
+   */
+  constructor(format) {
+    this[kType] = 'DecompressionStream';
+    switch (format) {
+      case 'deflate':
+        this[kHandle] = createInflate();
+        break;
+      case 'gzip':
+        this[kHandle] = createGunzip();
+        break;
+      default:
+        throw new ERR_INVALID_ARG_VALUE('format', format);
+    }
+    this[kTransform] = newReadableWritablePairFromDuplex(this[kHandle]);
+  }
+
+  /**
+   * @readonly
+   * @type {ReadableStream}
+   */
+  get readable() {
+    if (!isDecompressionStream(this))
+      throw new ERR_INVALID_THIS('DecompressionStream');
+    return this[kTransform].readable;
+  }
+
+  /**
+   * @readonly
+   * @type {WritableStream}
+   */
+  get writable() {
+    if (!isDecompressionStream(this))
+      throw new ERR_INVALID_THIS('DecompressionStream');
+    return this[kTransform].writable;
+  }
+
+  [kInspect](depth, options) {
+    if (!isDecompressionStream(this))
+      throw new ERR_INVALID_THIS('DecompressionStream');
+    customInspect(depth, options, 'DecompressionStream', {
+      readable: this[kTransform].readable,
+      writable: this[kTransform].writable,
+    });
+  }
+}
+
+ObjectDefineProperties(CompressionStream.prototype, {
+  readable: { enumerable: true },
+  writable: { enumerable: true },
+});
+
+ObjectDefineProperties(DecompressionStream.prototype, {
+  readable: { enumerable: true },
+  writable: { enumerable: true },
+});
```

Comment:
Would it be worth doing that? It's to avoid recreating the same object again and again, and be resilient if `Object.prototype.value` is set from userland.
```suggestion
const enumerableProperty = ObjectCreate(null);
enumerableProperty.enumerable = true;

ObjectDefineProperties(CompressionStream.prototype, {
  readable: enumerableProperty,
  writable: enumerableProperty,
});

ObjectDefineProperties(DecompressionStream.prototype, {
  readable: enumerableProperty,
  writable: enumerableProperty,
});
```

Reply:
I don't believe so. V8 is already going to optimize this part pretty well.

Reply:
And what about `Object.prototype` pollution?

Reply:
Ugh. Yeah, ok.

Reply:
See https://github.com/nodejs/node/pull/39371, once that lands I'll update this and the encoding streams PR to match.